feat: Integration map and CircleCI

SpainTrain opened the pull request:
- Added integration mapping for adding new CI integrations
- Added CircleCI integration

closes #5

### TODO

- [ ] Update README (add directions for CircleCI)

SpainTrain commented:
FYI, I am still testing this, but feel free to provide feedback on the approach

SpainTrain commented:
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/greenkeeperio/greenkeeper-shrinkwrap/7)*
<!-- Sent from Reviewable.io -->

boennemann commented:
Hey @SpainTrain,

I had a brief look at this and I'm super excited, as this is exactly what I had in mind.
Thanks for picking this up and executing it so exceptionally well.

I'll get to work on this again on Friday – I'll leave a proper review then.
Can't wait to get this merged soon.

Thanks a lot,
Stephan

SpainTrain commented:
Reviewed 2 of 2 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/greenkeeperio/greenkeeper-shrinkwrap/7)*
<!-- Sent from Reviewable.io -->

SpainTrain commented on lib/extract-dependency.js:
```diff
@@ -5,7 +5,7 @@ module.exports = function extractDependency (pkg, branch) {
   const branchPrefix = config.branchPrefix || 'greenkeeper/'
 
-  if (!branch.startsWith(branchPrefix)) throw new Error('Not a Greenkeeper pull request.')
+  if (!branch.startsWith(branchPrefix)) throw new Error('Not a Greenkeeper branch.')
```
I can change all these back, but I found it confusing (both as a user and dev) that it wasn't running on pull request builds, but then says in other messages that it has to run in a pull request build.

boennemann replied:
Totally agree, great change.

SpainTrain commented:
alright, was able to do some testing and looking solid.  I have *not* tested on Travis though.

SpainTrain commented:
I realized I also need to update the README with directions for CircleCI

boennemann commented on lib/ci-integrations/circle-ci.js:
```diff
@@ -0,0 +1,25 @@
+const env = process.env
+
+function test () {
+  return env.CIRCLECI === 'true'
```
I kinda like the idea to have a single file that contains all tests. We could then run the tests and only require the mapping of the CI service we're actually running on.

I know that the getValues function is only executed on demand, but that way we could avoid loading the mappings' dependencies into memory. We could then be more liberal about what individual CI services can `require` because it's not harming users on other platforms.

What do you think? Does it make sense, or is it an unnoticeable optimization?

boennemann commented on lib/ci-integrations/circle-ci.js:
```diff
@@ -0,0 +1,25 @@
+  }
+}
+
+module.exports = {
```
We're on min node 4, so we can make use of the shorthand. `module.exports = {test, getValues}`

Also feel free to make use of hoisting and move the export to the top of the file. I like seeing what's exported on top :)

boennemann commented on lib/get-values-from-ci.js:
```diff
@@ -0,0 +1,15 @@
+const integrationImports = [
+  require('./ci-integrations/travis-ci'),
+  require('./ci-integrations/circle-ci'),
+  {test: _.stubTrue, getValues: _.constant(null)}
```
I'd prefer arrow functions here. Less code. Less lodash specific knowledge required.
`{test: () => true, getValues: () => null}`

boennemann commented on lib/get-values-from-ci.js:
```diff
@@ -0,0 +1,15 @@
+]
+
+const integrations = _.map(integrationImports, function (integration) {
+  return [_.get(integration, 'test', _.stubFalse), _.get(integration, 'getValues', _.noop)]
```
I wouldn't use defaults here. If a ci-integration doesn't export these functions I'd prefer this to fail on us early.

boennemann commented on lib/get-values-from-ci.js:
```diff
@@ -0,0 +1,15 @@
+const _ = require('lodash')
+
+const integrationImports = [
+  require('./ci-integrations/travis-ci'),
```
Would love to have this filesystem based (`fs.readdirSync()`), as I hate manually maintaining such lists :)

boennemann replied:
Just realized, if we switch to having `test`s in a single file and requiring the `getValue`s on demand this is obsolete anyways.

boennemann commented on update.js:
```diff
@@ -3,38 +3,42 @@
   }
 
-  if (env.TRAVIS_PULL_REQUEST !== 'false') {
+  const isPullRequest = _.get(ciValues, 'isPullRequest', false)
```
Would be nice to have defaults applied in the lib file, so we can just use them here.
Also we don't need to repeat defaults if they're required in both upload and update.

boennemann commented:
Hey @SpainTrain,

I added some comments. Regarding the README – don't worry too much. If you don't want to spend time on it I can take care of if before merging – just let me know :)

Also: While using the current implementation I found a bug with the current approach of detecting whether the update was already applied. I think I found a super elegant solution for this on Travis, which I'd love to ship now, in order to unbreak current usage.
If I land this I'll link up the changes here, so we can look into rebasing this branch onto the changes.

Best,
Stephan

boennemann commented on update.js:
```diff
@@ -3,38 +3,42 @@
+  const ciValues = getValuesFromCI()
+
+  if (_.isEmpty(ciValues)) {
+    throw new Error('This script must be run in a supported CI environment')
```
`getValuesFromCI` could throw directly. Then we wouldn't need to copy this error message over to upload, too.

boennemann commented:
Alrighty, I just pushed the update. I found a lot more to refactor and given how much our branches diverged I started to feel bad for you here. I just pushed a branch where I've already resolved the merge conflicts so it's not too hard for you to pick this up again: https://github.com/greenkeeperio/greenkeeper-shrinkwrap/commits/integration-mapping-rebased

The most relevant change is this one: https://github.com/greenkeeperio/greenkeeper-shrinkwrap/commit/be944c0b56e9ffdd33941315ba6c01b28ba381b0

You would have to find an equivalent way of detecting this on CircleCI, and then add it to value mapping.

Best,
Stephan

SpainTrain commented:
Thanks for the review.  All good points.  I am currently slammed with work coding, but will hopefully get to rebasing and addressing comments in a week or so!

SpainTrain commented:
Haven't forgot about this, just slammed on other stuff.  Hopefully will be able to update quite soon!

boennemann commented:
Hey @SpainTrain,

thanks for your effort with this PR. I went ahead and added the mapping.
I'd be happy to have your Circle CI contribution: https://github.com/greenkeeperio/greenkeeper-lockfile#contributing-a-ci-service

Best,
Stephan
